src/linear_algebra/vandermonde.lean

```diff
@@ -0,0 +1,275 @@
+/-
+Copyright (c) 2021 Paul van Wamelen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul van Wamelen
+-/
+import linear_algebra.matrix
+import algebra.geom_sum
+
+/-!
+# Formula for the determinent of a Vandermonde matrix.
+
+## Theorems
+
+1. `vandermonde_det` gives the formula for the determinant of a Vandermonde matrix.
+
+## Implementation details
+
+The proof proceeds by three transformations of the matrix, each accomplished but multiplying by a
+matrix whose determinant can easily be written down. Transformation 1 is to subtract the first row
+from each of the others. This is multiplication by `vandermonde_step1_op`. After this
+transformation every entry below row one is the difference between two `i`th powers and allows us to
+factor `a i - a 0` from each row (except the first). This is multiplication by a diagonal matrix.
+The final step is to add the right multiples of each column to earlier columns in order to rewrite
+the matrix so the first minor is just a smaller Vandermonde matrix. This is multiplication by
+`vandermonde_step3_op`. The result now follows by induction.
+-/
+
+noncomputable theory
+open_locale classical
+
+universes u v
+variables (R : Type u) [comm_ring R] (n : ℕ) (a : ℕ → R)
```

Comment:
What's the thinking for having Vandermonde take an "infinite" sequence as input, does it make the proof simpler?
A version for functions from `fin n \to R` seems useful to add too?
As a separate statement even if you need this version as the main one?

Reply:
Yes, it helps a lot with the proof. But I agree it would be a good idea to have a `fin n → R` version also. I'll work on that.

Reply:
With `def vandermonde_fin (a : fin (n + 1) → R) : matrix (fin (n + 1)) (fin (n + 1)) R := λ i j, (a i) ^ (j : ℕ)`
which statement is best:
  `det (vandermonde_fin R n a) = ∏ i : fin (n + 1), ∏ j : {j : fin (n + 1) // j < i}, ((a i) - (a j))`
or
  `det (vandermonde_fin R n a) = ∏ i in range (n + 1), ∏ j in range i, ((a i) - (a j))`
?

Reply:
The prod over fin version seems better to me? I was also a little curious about the `+1`s everywhere, like if vandermonde was a matrix
```
matrix (fin n) (fin n) 
```
the statement
```
  det (vandermonde R n a) = ∏ i in range n, ∏ j in range i, ((a i) - (a j)) :=
```
would be true right?
Having this edge case of `n = 0` included might make your work easier to apply.

Reply:
The `n+1`s are needed for the definitions of all the other matrices (e.g. trying to define `vandermonde_step1` runs into `failed to synthesize type class instance for has_zero (fin n)`), but you are right, the result still holds for `n = 0`. I've made the change to include this case.
I also added the proof of 
`det (vandermonde_fin R n a) = ∏ i : fin (n + 1), ∏ j : {j : fin (n + 1) // j < i}, ((a i) - (a j))`